pyccl/halos/halo_model_base.py

```diff
@@ -0,0 +1,300 @@
+from .. import ccllib as lib
+from ..core import check
+from ..parameters import physical_constants as const
+from ..base import CCLAutoreprObject, warn_api, deprecated, deprecate_attr
+import numpy as np
+import functools
+from abc import abstractmethod
+
+
+__all__ = ("HMIngredients", "TinkerFunction",)
+
+
+def _subclasses(cls):
+    # This helper returns a set of all subclasses.
+    direct_subs = cls.__subclasses__()
+    deep_subs = [sub for cl in direct_subs for sub in cl._subclasses()]
+    return set(direct_subs).union(deep_subs)
+
+
+def from_name(cls, name):
+    """Obtain particular model."""
+    mod = {p.name: p for p in cls._subclasses() if hasattr(p, "name")}
+    return mod[name]
+
+
+def initialize_from_input(cls, input_, **kwargs):
+    """Process the input and generate an object of the class.
+    Input can be an instance of the class, or a name string.
+    Optional ``**kwargs`` may be passed.
+    """
```

Comment:
I know, I fully agree, just being lazy with the docstrings. As above.

Reply:
As far as the function's name, I don't like it either, I just had to come up with something quick.
This is a generalization of `_concentration_init` that used to be a method in `Concentration`.
I asked ChatGPT and it came out with the following names. Tell me which one you like better.

- `generate_object`
- `create_instance`
- `parse_input`
- `instantiate`

Reply:
I chose `create_instance` because it is clear and precise. Flows like natural language `MassDef.create_instance(**kwargs)`.

Re the docs, resolve as I've already taken care of that in 1053.